core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -93,7 +95,27 @@ protected void doExecute(SearchRequest searchRequest, ActionListener<SearchRespo
             logger.debug("failed to optimize search type, continue as normal", e);
         }
 
-        searchAsyncAction(searchRequest, listener).start();
+        ActionListener<SearchResponse> wrapper = listener;
+        // If a plugin has registered listeners for the search response, wrap the ActionListener
+        // and call each listener in turn.
+        if (!searchPhaseController.getSearchResponseListeners().isEmpty()) {
+            wrapper = new ActionListener<SearchResponse>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    for (BiConsumer<SearchRequest, SearchResponse> srl : searchPhaseController.getSearchResponseListeners()) {
```

Comment:
should we call the actual listener in a finally block? I really think we should otherwise we might not send a response to the client?